### PR TITLE
Fully-detailed output.

### PR DESCRIPTION
**Pull request by jamesmcclain**
The output of the program has been elaborated to include the individual
contributions of changed cells (caused by "modifications") within a
larger group of previously-indentical cells.

Test have been updated.

Connects to WikiWatershed/tr-55#29

### CODE REVIEW

**Comment by jamesmcclain** [comment]
The output now looks like this:
```Python
{u'modified': {u'bod': 1196.1951983335632,
               u'cell_count': 147,
               u'distribution': {u'a:deciduous_forest': {u'bod': 42.579624080381144,
                                                         u'cell_count': 72,
                                                         u'distribution': {u'a:deciduous_forest': {u'bod': 0.0,
                                                                                                   u'cell_count': 67,
                                                                                                   u'et': 26.516700000000018,
                                                                                                   u'inf': 34.9181,
                                                                                                   u'runoff': 0.0,
                                                                                                   u'tn': 0.0,
                                                                                                   u'tp': 0.0,
                                                                                                   u'tss': 0.0},
                                                                           u'd:rock': {u'bod': 42.579624080381144,
                                                                                       u'cell_count': 5,
                                                                                       u'et': 0.0,
                                                                                       u'inf': 29.360978816938722,
                                                                                       u'runoff': 11.519021183061302,
                                                                                       u'tn': 0.0032257290969985716,
                                                                                       u'tp': 0.0003225729096998572,
                                                                                       u'tss': 0.32257290969985714}},
                                                         u'et': 24.675262500000017,
                                                         u'inf': 34.53218880673186,
                                                         u'runoff': 0.7999320266014792,
                                                         u'tn': 0.0032257290969985716,
                                                         u'tp': 0.0003225729096998572,
                                                         u'tss': 0.32257290969985714},
                                 u'c:commercial': {u'bod': 619.3321622220828,
                                                   u'cell_count': 42,
                                                   u'distribution': {u'c:commercial': {u'bod': 555.7691766958419,
                                                                                       u'cell_count': 22,
                                                                                       u'et': 2.2728600000000068,
                                                                                       u'inf': 4.430079770137537,
                                                                                       u'runoff': 36.37540022986247,
                                                                                       u'tn': 4.078628635429163,
                                                                                       u'tp': 0.6454093664854939,
                                                                                       u'tss': 113.35001998901487},
                                                                     u'c:no_till': {u'bod': 63.56298552624096,
                                                                                    u'cell_count': 20,
                                                                                    u'et': 34.0929,
                                                                                    u'inf': 29.143644157107012,
                                                                                    u'runoff': 4.576255842892978,
                                                                                    u'tn': 0.4664702970070909,
                                                                                    u'tp': 0.07381507996595724,
                                                                                    u'tss': 12.96377341902124}},
                                                   u'et': 17.42526,
                                                   u'inf': 16.198443763932524,
                                                   u'runoff': 21.23295052178176,
                                                   u'tn': 4.545098932436254,
                                                   u'tp': 0.7192244464514511,
                                                   u'tss': 126.3137934080361},
                                 u'd:hi_residential': {u'bod': 534.2834120310993,
                                                       u'cell_count': 33,
                                                       u'distribution': {u'd:hi_residential': {u'bod': 488.9532547665262,
                                                                                               u'cell_count': 23,
                                                                                               u'et': 6.818579999999968,
                                                                                               u'inf': 8.361629213022571,
                                                                                               u'runoff': 32.167342828759615,
                                                                                               u'tn': 2.83841508063619,
                                                                                               u'tp': 0.47652223981483477,
                                                                                               u'tss': 58.425770272949315},
                                                                         u'd:no_till': {u'bod': 45.3301572645731,
                                                                                        u'cell_count': 10,
                                                                                        u'et': 34.0929,
                                                                                        u'inf': 26.860866751950642,
                                                                                        u'runoff': 6.859033248049366,
                                                                                        u'tn': 0.26314540445959805,
                                                                                        u'tp': 0.0441776956392026,
                                                                                        u'tss': 5.416569639241362}},
                                                       u'et': 15.08352545454543,
                                                       u'inf': 13.967458770273502,
                                                       u'runoff': 24.498158107332266,
                                                       u'tn': 3.101560485095788,
                                                       u'tp': 0.5206999354540374,
                                                       u'tss': 63.842339912190674}},
               u'et': 20.450586122448982,
               u'inf': 24.67740388835977,
               u'runoff': 11.957947247429287,
               u'tn': 7.649885146629041,
               u'tp': 1.2402469548151882,
               u'tss': 190.47870622992664},
 u'unmodified': {u'bod': 1762.555509187117,
                 u'cell_count': 147,
                 u'distribution': {u'a:deciduous_forest': {u'bod': 0.0,
                                                           u'cell_count': 72,
                                                           u'et': 26.51670000000007,
                                                           u'inf': 34.91810000000001,
                                                           u'runoff': 0.0,
                                                           u'tn': 0.0,
                                                           u'tp': 0.0,
                                                           u'tss': 0.0},
                                   u'c:commercial': {u'bod': 1061.0138827829708,
                                                     u'cell_count': 42,
                                                     u'et': 2.272860000000005,
                                                     u'inf': 4.430079770137537,
                                                     u'runoff': 36.375400229862464,
                                                     u'tn': 7.786472849455671,
                                                     u'tp': 1.2321451541995787,
                                                     u'tss': 216.39549270630107},
                                   u'd:hi_residential': {u'bod': 701.5416264041463,
                                                         u'cell_count': 33,
                                                         u'et': 6.818579999999993,
                                                         u'inf': 8.361629213022574,
                                                         u'runoff': 32.167342828759615,
                                                         u'tn': 4.072508593956273,
                                                         u'tp': 0.6837058223430238,
                                                         u'tss': 83.8282790872751}},
                 u'et': 15.167861632653095,
                 u'inf': 20.24558036990151,
                 u'runoff': 17.614211721110824,
                 u'tn': 11.858981443411944,
                 u'tp': 1.9158509765426026,
                 u'tss': 300.2237717935762}}
```
This tree-like dictionary breaks down the contributions of each cell type, as well modifications to bunches of cell types.

File: test/test_model.py [hunk 1]
```diff
@@ -23,583 +24,6 @@
 CN70 = [0.000, 0.030, 0.060, 0.110, 0.170, 0.240, 0.460, 0.710, 1.010, 1.330, 1.670, 2.040, 2.810, 3.620, 4.460, 5.330, 6.220, 7.130, 8.050, 8.980, 9.910, 10.85]  # noqa
 CN80 = [0.080, 0.150, 0.240, 0.340, 0.440, 0.560, 0.890, 1.250, 1.640, 2.040, 2.460, 2.890, 3.780, 4.690, 5.630, 6.570, 7.520, 8.480, 9.450, 10.42, 11.39, 12.37]  # noqa
 CN90 = [0.320, 0.460, 0.610, 0.760, 0.930, 1.090, 1.530, 1.980, 2.450, 2.920, 3.400, 3.880, 4.850, 5.820, 6.810, 7.790, 8.780, 9.770, 10.76, 11.76, 12.75, 13.74]  # noqa
-# INPUT and OUTPUT are data that were emailed to Azavea in a
-# spreadsheet for testing the TR-55 model implementation.
-INPUT = [
```
Comment (mmcfarland) [inline test/test_model.py]: This seemed like a good test to match our previous implementation, was it necessary to remove?

**Comment by jamesmcclain** [comment]
One more note about this: this code does produce negative runoff values under some scenarios when the PItt Small Storm method is used.  It is my recollection that I was aware of this and intentionally allowed it because I saw the same behaviour in the Scala code (although I cannot be absolutely certain of that right now, because I am not looking at the Scala implementation).

If that is problematic, I would be more than happy to fix it (a one line fix).

File: test/test_model.py (hunk 1 above)
Comment (jamesmcclain) [inline test/test_model.py]: No, it didn't need to be removed, but it didn't seem like it was proving much.  Essentially, it was just showing that the output was similar to that from the spreadsheet.  In other words, the numeric bounds were essentially just arbitrary numbers.

**Comment by mmcfarland** [comment]
Output looks good, though it's dense enough that it could warrant some explanation in the readme.  In the following example, I'm assuming that in the unmodified cell census there were 33 cells of `d:hi_residential`, of which 10 cells were modified to use a BMP `no_till`, and we can do a little subtraction from the unmodified branch of the tree to find out the overall change an individual modification reflected, plus the contributions of sub regions.  
screenshot from 2015-05-14 10 32 23

**Comment by jamesmcclain** [comment]
Okay, sounds good.  I will add some discussion to the `README.md` file.

**Comment by jamesmcclain** [comment]
I have added some information to the `README.md` file about how to interpret the output.

**Comment by mmcfarland** [comment]
+1 to merge

**Comment by jamesmcclain** [comment]
Thank you.

**Pull request merged**